### PR TITLE
Compile model+loss for LoRA single device recipe

### PR DESCRIPTION
**Pull request by gau-nernst**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

Address #1228

To limit the scope of this PR, I will only update the LoRA single device recipe. We can keep #1228 open and discuss how to do model+loss compile for other recipes.

#### Changelog
What are the changes made in this PR?

Compile model+loss together for LoRA single device recipe. This has shown substantial speed improvements.

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](../CONTRIBUTING.md) for some guidance on contributing.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](../tests/torchtune) for any new functionality
- [ ] update [docstrings](../docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

Run the following command on 4070Ti SUPER (16GB VRAM)

>`tune run lora_finetune_single_device --config llama3/8B_qlora_single_device compile=True metric_logger._component_=torchtune.utils.metric_logging.WandBLogger batch_size=4 gradient_accumulation_steps=1 log_peak_memory_stats=True`

image

15% improvement in tok/s on average, generally lower memory consumption.

### CODE REVIEW

File: recipes/lora_finetune_single_device.py [hunk 1]
```diff
@@ -557,31 +576,10 @@ def train(self) -> None:
...
-                    logits = self._model(tokens, mask=mask, input_pos=input_pos)
-                    # Shift so that tokens < n predict n
-                    logits = logits[..., :-1, :].contiguous()
-                    labels = labels[..., 1:].contiguous()
-                    logits = logits.transpose(1, 2)
-                    # Compute loss
-                    loss = self._loss_fn(logits, labels)
-                    # free logits otherwise it peaks backward memory
-                    del logits
+                    batch = {k: v.to(self._device) for k, v in batch.items()}
```
Comment (felipemello1) [inline recipes/lora_finetune_single_device.py:579]: My concern is that the batch may contain keys that may not be necessary in the gpu. Can we keep the .to(device) like it was before? Or did you have some other motivation for it?

Comment (gau-nernst) [reply recipes/lora_finetune_single_device.py:579]: I tried putting the old `.to(device)` code within `self._loss_step()`, but `torch.compile()` shows some warnings. Should be harmless though. I will double-check it again.

Another option is to keep the old `.to(device)` outside `self._loss_step()`. Then, it is a bit more clunky to pass `tokens`, `labels`, `mask`, `input_pos` to `self._loss_step()` instead of just `batch` (or we can just modify the `batch` object in-place). Lmk which one you prefer, then I will just change to that.

Comment (felipemello1) [reply recipes/lora_finetune_single_device.py:579]: i think it should be fine to keep as it is then. I dont think that our batches output extra terms, so that scenario I raised would be an exception.

**Comment by felipemello1** [comment]
Before i approve it, would it make sense to add it to the other recipes, so they are all aligned? For example, should we test it in distributed? We probably want to avoid having different versions of loss_step for each recipe.

out of curiosity: Did you notice any major difference in compiling time?

**Comment by gau-nernst** [comment]
> would it make sense to add it to the other recipes, so they are all aligned?

I think we can add this to other recipes in separate PR, since each requires good testing to confirm no regression and expected speed up. For example, even for this recipe (single device QLoRA), took me 10min+ for each mini run with Lllama3-8B (well most of the time is spent on quantization due to #1246 and compilation).

> should we test it in distributed

I don't have a multi-gpu setup, so I can't run and test it myself 😅.

> We probably want to avoid having different versions of loss_step for each recipe.

I think the point is that `self._loss_step()` will be different for each recipe? so that each recipe can have their custom logic. But yea at least among the ones that trains on autogressive modelling loss (not DPO/PPO), their `self._loss_step()` should be the same.

> Did you notice any major difference in compiling time?

Not on my machine. I also checked it myself. Didn't put a timer, but this wandb graph suggests they have the same compile time (power usage goes up at the same time -> training starts at the same time)

image

Another question. Realised that I deleted this line of yours

https://github.com/pytorch/torchtune/blob/c1fabb958873f5356133e897b66b889a6477de57/recipes/full_finetune_single_device.py#L523

Since now we compile model+loss together, I don't think this will make a difference for `compile=True` anymore. But probably still has an impact in eager mode (`compile=False`). I will add it back. Is it ok with you?

**Comment by felipemello1** [comment]
sounds good! gonna let the tests run and approve the PR. Thanks for addressing my questions

**Review by felipemello1 — APPROVED** [review]
